core/src/test/java/com/linecorp/armeria/internal/common/stream/PrependingPublisherTckTest.java

```diff
@@ -0,0 +1,135 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.stream;
+
+import java.util.stream.LongStream;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import org.reactivestreams.tck.PublisherVerification;
+import org.reactivestreams.tck.TestEnvironment;
+import org.reactivestreams.tck.flow.support.PublisherVerificationRules;
+import org.testng.SkipException;
+import org.testng.annotations.Test;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+@SuppressWarnings("checkstyle:LineLength")
+public class PrependingPublisherTckTest extends PublisherVerification<Object> {
+
+    public PrependingPublisherTckTest() {
+        super(new TestEnvironment(200));
+    }
+
+    @Override
+    public Publisher<Object> createPublisher(long elements) {
+        if (elements == 0) {
+            return Mono.empty();
+        }
+        return new PrependingPublisher<>("Hello", Flux.fromStream(LongStream.range(0, elements - 1).boxed()));
+    }
+
+    /**
+     * Rule 1.4 and 1.9 ensure a Publisher's ability to signal error to the Subscriber, however the
+     * implementation expects such error to occur immediately after subscribing, i.e. {@code onError()} is
+     * called after {@code onSubscribe()}. The {@link PrependingPublisher} however always serves at least one
+     * element before failing, therefore for the error to be signaled, we must make requests first.
+     *
+     * {@link PublisherVerificationRules#optional_spec104_mustSignalOnErrorWhenFails()} and
+     * {@link PublisherVerificationRules#required_spec109_mayRejectCallsToSubscribeIfPublisherIsUnableOrUnwillingToServeThemRejectionMustTriggerOnErrorAfterOnSubscribe()}
+     * are overridden below to call {@link Subscription#request(long)} after subscribing.
+     */
+    @Override
+    public Publisher<Object> createFailedPublisher() {
+        return new PrependingPublisher<>("Hello", Mono.error(new RuntimeException()));
+    }
+
+    @Test
+    @Override
+    public void optional_spec104_mustSignalOnErrorWhenFails() {
+        try {
+            final TestEnvironment env = new TestEnvironment(200);
+            whenHasErrorPublisherTest(pub -> {
+                final TestEnvironment.Latch onErrorLatch = new TestEnvironment.Latch(env);
+                final TestEnvironment.Latch onSubscribeLatch = new TestEnvironment.Latch(env);
+                pub.subscribe(new TestEnvironment.TestSubscriber<Object>(env) {
+                    @Override
+                    public void onSubscribe(Subscription subs) {
+                        onSubscribeLatch.assertOpen("Only one onSubscribe call expected");
+                        onSubscribeLatch.close();
+                        subs.request(Long.MAX_VALUE);
+                    }
+
+                    @Override
+                    public void onNext(Object element) {
```

Comment:
How about adding an assertion for always receiving the first element from the failed publisher?